Frameworks/CoreGraphics/CGContextCairo.mm

```diff
@@ -1633,6 +1634,358 @@
+
+        double miterLimit = cairo_get_miter_limit(context);
+        cairo_line_join_t lineJoin = cairo_get_line_join(context);
+
+        // Check to see if join should be overridden to use bevel
+        CGFloat delta = angle2 - angle1;
+        delta = fmod(delta, 2.0f * M_PI);
+        if (delta < 0.0f) {
+            delta += 2.0f * M_PI;
+        }
+        // Force acute angles to use bevel
+        // Top range is temporary fix to avoid joins
+        // on intermediate curve lines
+        if (delta < M_PI || delta > (1.98f * M_PI)) {
+            lineJoin = CAIRO_LINE_JOIN_BEVEL;
+        }
+
+        if (lineJoin == CAIRO_LINE_JOIN_MITER) {
+            CGFloat m1 = tan(angle1);
+            CGFloat b1 = p1.y - (m1 * p1.x);
+
+            CGFloat m2 = tan(angle2);
+            CGFloat b2 = p2.y - (m2 * p2.x);
+
+            CGFloat x = (b2 - b1) / (m1 - m2);
+            CGFloat y = 0;
+            // If m1 is very large it's probably a vertical line
+            if (m1 > 1000000) {
+                y = m2 * x + b2;
+            } else {
+                y = m1 * x + b1;
+            }
+            CGFloat fullStrokeWidth = 2.0f * halfStrokeWidth;
+            CGPoint miterp1 = strokedPoint(p1, angle1, fullStrokeWidth, !reverse);
+            CGPoint miterp2 = strokedPoint(p2, angle2, fullStrokeWidth, !reverse);
+
+            CGFloat miterm1 = tan(angle1);
+            CGFloat miterb1 = miterp1.y - (miterm1 * miterp1.x);
+
+            CGFloat miterm2 = tan(angle2);
+            CGFloat miterb2 = miterp2.y - (miterm2 * miterp2.x);
+
+            CGFloat miterx = (miterb2 - miterb1) / (miterm1 - miterm2);
+            CGFloat mitery = 0;
+            // If m1 is very large it's probably a vertical line
+            if (miterm1 > 1000000) {
+                mitery = miterm2 * miterx + miterm2;
+            } else {
+                mitery = miterm1 * miterx + miterb1;
+            }
+
+            CGFloat miterLength = sqrt(pow(miterx - x, 2.0f) + pow(mitery - y, 2.0f));
+
+            CGFloat miterRatio = (miterLength / fullStrokeWidth);
+
+            if (miterRatio > miterLimit) {
+                lineJoin = CAIRO_LINE_JOIN_BEVEL;
+            } else {
+                cairo_line_to(context, x, y);
+                cairo_line_to(context, p2.x, p2.y);
+                return;
+            }
+        }
+
+        if (lineJoin == CAIRO_LINE_JOIN_ROUND) {
+            CGFloat m1 = tan(angle1);
+            CGFloat b1 = p1.y - (m1 * p1.x);
+
+            CGFloat m2 = tan(angle2);
+            CGFloat b2 = p2.y - (m2 * p2.x);
+
+            CGFloat x = (b2 - b1) / (m1 - m2);
+
+            CGFloat y = 0;
+            // If m1 is very large it's probably a vertical line
+            if (m1 > 1000000) {
+                y = m2 * x + b2;
+            } else {
+                y = m1 * x + b1;
+            }
+
+            // TODO: This code comes directly from CGContextAddArcToPoint
+            // Consolidate and reuse instead of copy
```

Comment:
Yes, this code is copied a couple different spots, but it also needs refactoring/commenting to make it clear what it's doing. The other spot it was copied to has a similar todo comment.